File: app/src/main/java/com/semisonfire/cloudgallery/ui/main/trash/TrashPresenter.java
package com.semisonfire.cloudgallery.ui.main.trash;

import android.text.TextUtils;

import com.semisonfire.cloudgallery.data.local.prefs.DiskPreferences;
import com.semisonfire.cloudgallery.data.model.Photo;
import com.semisonfire.cloudgallery.data.remote.RemoteDataSource;
import com.semisonfire.cloudgallery.data.remote.api.DiskClient;
import com.semisonfire.cloudgallery.ui.base.BasePresenter;

import java.util.ArrayList;
import java.util.List;

import io.reactivex.Flowable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;

public class TrashPresenter<V extends TrashContract.View> extends BasePresenter<V>
        implements TrashContract.Presenter<V> {

    private RemoteDataSource mRemoteDataSource;

    public TrashPresenter(DiskPreferences preferences, RemoteDataSource remoteDataSource) {
        super(preferences);
        mRemoteDataSource = remoteDataSource;
    }

    @Override
    public void getPhotos(int page) {
        getCompositeDisposable().add(
                mRemoteDataSource.getTrashPhotos(DiskClient.MAX_LIMIT, page)
                        .flatMapIterable(photos -> photos)
                        .filter(photo -> !TextUtils.isEmpty(photo.getMediaType()))
                        .filter(photo -> photo.getMediaType().equals("image"))
                        .toList()
                        .subscribeOn(Schedulers.io())
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribe(photos -> getMvpView().onTrashLoaded(photos),
                                throwable -> getMvpView().onError(throwable)));
    }

    @Override
    public void restorePhotos(List<Photo> photos) {
        List<Photo> items = new ArrayList<>(photos);
        getCompositeDisposable().add(
                Flowable.fromIterable(items)
                        .concatMap(photo -> mRemoteDataSource.restoreTrashPhoto(photo))
                        .subscribeOn(Schedulers.io())
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribe(p -> getMvpView().onPhotoRestored(p),
                                throwable -> getMvpView().onError(throwable)));
    }

    @Override
    public void deletePhotos(List<Photo> photos) {
        List<Photo> items = new ArrayList<>(photos);
        getCompositeDisposable().add(
                Flowable.fromIterable(items)
                        .concatMap(photo -> mRemoteDataSource.deleteTrashPhoto(photo))
                        .subscribeOn(Schedulers.io())
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribe(p -> getMvpView().onPhotoDeleted(p),
                                throwable -> getMvpView().onError(throwable)));
    }

    @Override
    public void clear() {
        getCompositeDisposable().add(
                mRemoteDataSource.clearTrash()
                        .subscribeOn(Schedulers.io())
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribe(() -> getMvpView().onTrashCleared(),
                                throwable -> getMvpView().onError(throwable)));
    }

}

File: app/src/main/java/com/semisonfire/cloudgallery/ui/trash/TrashPresenter.kt
package com.semisonfire.cloudgallery.ui.trash

import com.semisonfire.cloudgallery.core.presentation.BasePresenter
import com.semisonfire.cloudgallery.data.model.Photo
import com.semisonfire.cloudgallery.data.remote.RemoteRepository
import com.semisonfire.cloudgallery.ui.disk.LIMIT
import com.semisonfire.cloudgallery.utils.background
import com.semisonfire.cloudgallery.utils.foreground
import com.semisonfire.cloudgallery.utils.printThrowable
import io.reactivex.Flowable
import java.util.*

class TrashPresenter(private val remoteRepository: RemoteRepository) :
  BasePresenter<TrashContract.View>(), TrashContract.Presenter {

  override fun getPhotos(page: Int) {
    compositeDisposable.add(
      remoteRepository.getTrashPhotos(LIMIT, page)
        .flatMapIterable { it }
        .filter { it.mediaType.isNotEmpty() }
        .filter { it.mediaType == "image" }
        .toList()
        .subscribeOn(background())
        .observeOn(foreground())
        .subscribe(
          { view?.onTrashLoaded(it) },
          { it.printThrowable() }
        )
    )
  }

  override fun restorePhotos(photos: List<Photo>) {
    val items: List<Photo> = ArrayList(photos)
    compositeDisposable.add(
      Flowable.fromIterable(items)
        .concatMap { remoteRepository.restoreTrashPhoto(it) }
        .subscribeOn(background())
        .observeOn(foreground())
        .subscribe(
          { view?.onPhotoRestored(it) },
          { it.printThrowable() }
        )
    )
  }

  override fun deletePhotos(photos: List<Photo>) {
    val items: List<Photo> = ArrayList(photos)
    compositeDisposable.add(
      Flowable.fromIterable(items)
        .concatMap { remoteRepository.deleteTrashPhoto(it) }
        .subscribeOn(background())
        .observeOn(foreground())
        .subscribe(
          { view?.onPhotoDeleted(it) },
          { it.printThrowable() }
        )
    )
  }

  override fun clear() {
    compositeDisposable.add(
      remoteRepository.clearTrash()
        .subscribeOn(background())
        .observeOn(foreground())
        .subscribe(
          { view?.onTrashCleared() },
          { it.printThrowable() }
        )
    )
  }
}